File: src/databases/key_value/key_value.rs
use std::collections::HashMap;

use crate::structs::errors::KVError;
pub trait KeyValue {
    fn connect(self) -> Result<bool, KVError>;
    fn disconnect(self) -> Result<bool, KVError>;

    fn ping(self) -> Result<u64, KVError>;

    fn get(self, key: String) -> Result<String, KVError>;

    fn set_s(self, key: String, value: String) -> Result<bool, KVError>;
    fn set_u64(self, key: String, value: u64) -> Result<bool, KVError>;

    fn set_s_ex(self, key: String, value: String, expire: u64) -> Result<bool, KVError>;
    fn set_u64_ex(self, key: String, value: u64, expire: u64) -> Result<bool, KVError>;

    fn exists(self, key: String) -> Result<bool, KVError>;

    fn incr(self, key: String) -> Result<bool, KVError>;
    fn incr_by(self, key: String, incr: u64) -> Result<bool, KVError>;

    fn delete(self, key: String) -> Result<bool, KVError>;

    fn add_mul_s(self, kv_set: HashMap<String, String>) -> Result<bool, KVError>;
    fn add_mul_u64(self, kv_set: HashMap<String, u64>) -> Result<bool, KVError>;

    fn get_mul(self, keys: Vec<String>) -> Result<HashMap<String, String>, KVError>;

    fn get_type(self, key: String) -> Result<String, KVError>;

    fn set_ex(self, key: String, expire: u64) -> Result<bool, KVError>;
}

File: src/main/java/com/alexcomeau/database/keyvalue/KeyValue.java
package com.alexcomeau.database.keyvalue;

import java.util.ArrayList;
import java.util.HashMap;

import com.alexcomeau.database.DatabaseException;

public interface KeyValue {
    public void Connect() throws DatabaseException;

    public void Disconnect() throws DatabaseException;

    public String get(String key) throws DatabaseException;
    
    public void set(String key, String value) throws DatabaseException;

    public void setKeyExpire(String key, long exp) throws DatabaseException;

    public void setWExpire(String key, String value, long expire) throws DatabaseException;

    public boolean exists(String key) throws DatabaseException;

    public long incr(String key) throws DatabaseException;
    public long incrBy(String key, long incr) throws DatabaseException;

    public void delete(String key) throws DatabaseException;

    public void addMultiple(HashMap<String, String> hMap) throws DatabaseException;

    public HashMap<String, String> getMultiple(ArrayList<String> get) throws DatabaseException;
    
    public String getType(String key) throws DatabaseException;
}
